### PR TITLE
Allow to save unparsed records and errors in RabbitMQ, NATS and FileLog engines

### LINKED ISSUE #36035: Fallback format and table for streaming engines (Kafka, RabbitMQ, FileLog)
**Use case**

Put unparsed records into a separate table. The records will be splitted by fallback format, like LineAsString or RawBLOB.

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow to save unparsed records and errors in RabbitMQ, NATS and FileLog engines. Add virtual columns `_error` and `_raw_message`(for NATS and RabbitMQ), `_raw_record` (for FileLog) that are filled when ClickHouse fails to parse new record. The behaviour is controlled under storage settings `nats_handle_error_mode` for NATS, `rabbitmq_handle_error_mode` for RabbitMQ, `handle_error_mode` for FileLog similar to `kafka_handle_error_mode`. If it's set to `default`, en exception will be thrown when ClickHouse fails to parse a record, if it's set to `stream`, erorr and raw record will be saved into virtual columns. Closes https://github.com/ClickHouse/ClickHouse/issues/36035

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 9ca92c1b91ea8bcd5b877f0371c7fb8585a05249 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55477/9ca92c1b91ea8bcd5b877f0371c7fb8585a05249/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>❌ failure</td></tr>
<tbody>
</table>

File: src/Storages/FileLog/FileLogSource.cpp [hunk 1]
```diff
@@ -77,29 +79,72 @@ Chunk FileLogSource::generate()
...
+                // before exception, we need to fix it.
+                auto cur_rows = column->size();
+                if (cur_rows > total_rows)
+                    column->popBack(cur_rows - total_rows);
+
+                // All data columns will get default value in case of error.
+                column->insertDefault();
+            }
+
+            return 1;
```
Comment (kssenii) [inline src/Storages/FileLog/FileLogSource.cpp:106]: return value seems not needed

File: src/Storages/FileLog/FileLogConsumer.cpp [hunk 2]
```diff
@@ -155,21 +152,14 @@ void ReadBufferFromFileLog::readNewRecords(ReadBufferFromFileLog::Records & new_
     }
 }
 
-bool ReadBufferFromFileLog::nextImpl()
+ReadBufferPtr FileLogConsumer::getNextRecord()
 {
-    if (!allowed || !hasMorePolledRecords())
```
Comment (kssenii) [inline src/Storages/FileLog/FileLogConsumer.cpp:160]: `allowed` can probably be deleted now from this class, seems it is unused after this change

**Review by kssenii — APPROVED** [review]

**Comment by deevil** [comment]
For error messages length(_error) > 0 is correct.

But for good messages check length(_error) = 0 not working.
length(_error) is equal NULL

**Comment by Avogar** [comment]
For RabbitMQ, NATS and FileLog I made columns `_error` and ` _raw_record ` `Nullable`, it's more natural., In Kafka I didn't change it for compatibility reasons. You should check if there is an error by checking `_error is Null`

**Comment by deevil** [comment]
@Avogar it's cool.
But not documented(

**Comment by Avogar** [comment]
> But not documented(

Thanks for noticing it, I will add this to documentation

**Pull request merged**